internal/compiler/passes/embed_images.rs

```diff
@@ -10,27 +10,57 @@ use crate::EmbedResourcesKind;
 use image::GenericImageView;
 use std::cell::RefCell;
 use std::collections::HashMap;
+use std::future::Future;
+use std::pin::Pin;
 use std::rc::Rc;
 
-pub fn embed_images(
+pub async fn embed_images(
     component: &Rc<Component>,
     embed_files: EmbedResourcesKind,
     scale_factor: f64,
+    resource_url_mapper: &Option<Rc<dyn Fn(&str) -> Pin<Box<dyn Future<Output = Option<String>>>>>>,
     diag: &mut BuildDiagnostics,
 ) {
     let global_embedded_resources = &component.embedded_file_resources;
 
-    for component in component
+    let all_components = component
         .used_types
         .borrow()
         .sub_components
         .iter()
         .chain(component.used_types.borrow().globals.iter())
         .chain(std::iter::once(component))
-    {
+        .cloned()
+        .collect::<Vec<_>>();
+
+    // Collect URLs (sync!):
+    let urls = {
+        let mut urls = HashMap::<String, Option<String>>::new();
+        for component in &all_components {
+            visit_all_expressions(component, |e, _| {
+                collect_image_urls_from_expression(e, &mut urls)
+            });
+        }
+        urls
+    };
+
+    // Map URLs (async -- well, not really):
```

Comment:
load_dependencies_recursively_impl  has an example on how to make it really async.

But don't spend time on that, it can also be done later